lib/phoenix/controller.ex

```diff
@@ -405,26 +405,32 @@ defmodule Phoenix.Controller do
 
   defp url(opts) do
     cond do
-      to = opts[:to] -> validate_local_url(to)
+      to = opts[:to] -> validate_redirect_to(to)
       external = opts[:external] -> external
       true -> raise ArgumentError, "expected :to or :external option in redirect/2"
     end
   end
-  @invalid_local_url_chars ["\\"]
-  defp validate_local_url("//" <> _ = to), do: raise_invalid_url(to)
-  defp validate_local_url("/" <> _ = to) do
-    if String.contains?(to, @invalid_local_url_chars) do
-      raise ArgumentError, "unsafe characters detected for local redirect in URL #{inspect to}"
-    else
-      to
+
+  defp validate_redirect_to(to) do
+    case validate_local_url(to) do
+      {:ok, to} -> to
+      {:error, :invalid_local_url} -> raise ArgumentError, "the :to option in redirect expects a path but was #{inspect to}"
+      {:error, :unsafe_chars} -> raise ArgumentError, "unsafe characters detected for local redirect in URL #{inspect to}"
     end
   end
-  defp validate_local_url(to), do: raise_invalid_url(to)
 
-  @spec raise_invalid_url(term()) :: no_return()
-  defp raise_invalid_url(url) do
-    raise ArgumentError, "the :to option in redirect expects a path but was #{inspect url}"
+  @invalid_local_url_chars ["\\"]
+  @doc false
+  @spec validate_local_url(String.t) :: {:ok, String.t} | {:error, :invalid_local_url | :unsafe_chars}
+  def validate_local_url("//" <> _), do: {:error, :invalid_local_url}
```

Comment:
This means this function will end-up in all controllers since they do import `Phoenix.Controller`. I would say a little bit of code duplication is fine in this case, definitely better than the coupling, just add the logic directly to Phoenix.Endpoint and we should be golden!